### PR TITLE
Import individual Youtube videos via ID

### PR DESCRIPTION
**Pull request by delicioustoast**
Will write tests later.

Go to /videos/new to import a new video
Youtube ID and subject are mandatory fields

Reviewer: @DarryQueen @nalnaji @jonathanshum

### CODE REVIEW

**Comment by DarryQueen** [comment]
See https://github.com/DarryQueen/Khan-Burmese/issues/97 regarding the issue with the Travis build.

**Comment by delicioustoast** [comment]
I think I pushed to a branch on the main repo instead of my fork.

**Comment by DarryQueen** [comment]
You have a different error. See Travis build.

**Comment by DarryQueen** [comment]
Good start. I'll take a look in the morning.

File: app/controllers/videos_controller.rb [hunk 1]
```diff
@@ -30,15 +30,33 @@ def toggle_star
...
-    if request.post?
-      begin
-        Video.import(params[:file])
-        add_flash(:notice, 'Video(s) imported!')
-        redirect_to videos_path
-      rescue ArgumentError => e
-        add_flash(:alert, "#{e.message}")
-        redirect_to import_videos_path
-      end
+    return unless request.post?
```
Comment (DarryQueen) [inline app/controllers/videos_controller.rb]: Cute refactoring, but kind of belongs in its own commit.

File: app/models/video.rb [hunk 2]
```diff
@@ -64,13 +64,14 @@ def fill_missing_fields
 
     begin
       youtube_values = YoutubeReader::parse_video(self.youtube_id)
-    rescue
+    rescue SocketError, NoMethodError
+      raise ArgumentError, "Invalid Youtube ID"
```
Comment (DarryQueen) [inline app/models/video.rb]: You'll want to add punctuation to your errors since these messages get rendered directly into the errors bucket.

File: app/models/video.rb [hunk 3]
```diff
@@ -2,7 +2,7 @@ class Video < ActiveRecord::Base
   require 'YoutubeReader'
   require 'net/http'
 
-  attr_accessible :description, :title, :youtube_id, :starred, :duration
+  attr_accessible :description, :title, :youtube_id, :starred, :duration, :subject
```
Comment (DarryQueen) [inline app/models/video.rb]: Subject isn't per se an accessible attribute, since it's an ActsAsTaggableOn model.

File: app/models/video.rb (hunk 2 above)
Comment (DarryQueen) [inline app/models/video.rb]: On another note, this error probably belongs in the YoutubeReader module.

File: app/models/video.rb [hunk 4]
```diff
@@ -64,13 +64,14 @@ def fill_missing_fields
...
-    rescue
+    rescue SocketError, NoMethodError
+      raise ArgumentError, "Invalid Youtube ID"
     end
 
     unless youtube_values.empty?
       fields = [ 'description', 'title', 'duration' ]
       fields.each do |field|
-        write_attribute(field, youtube_values[field]) unless self[field]
+        write_attribute(field, youtube_values[field]) if self[field].to_s.empty?
```
Comment (DarryQueen) [inline app/models/video.rb]: `.blank?` is a nice way to check if an object is nil or empty.

File: app/models/video.rb [hunk 5]
```diff
@@ -136,13 +137,30 @@ def self.import(file)
       attributes = row.to_hash.slice(*keys)
 
       video = Video.new(attributes)
-      video.subject_list.add(row['subject'])
```
Comment (DarryQueen) [inline app/models/video.rb]: Can you justify this change? What was wrong with it before?

File: app/models/video.rb [hunk 6]
```diff
@@ -136,13 +137,30 @@ def self.import(file)
...
+        video.subject_list.add(row['subject'].strip.downcase)
+      end
 
       video.fill_missing_fields
       video.save
     end
   end
 
+  def self.import_id(video_hash)
+    if video_hash["youtube_id"].empty? or video_hash["subject"].empty?
```
Comment (DarryQueen) [inline app/models/video.rb]: Unless you need to escape inside your string, I would recommend single quotes.

File: app/models/video.rb [hunk 7]
```diff
@@ -136,13 +137,30 @@ def self.import(file)
...
+      raise ArgumentError, "Please fill in required fields."
+    end
+
+    keys = [ 'description', 'title', 'youtube_id']
+    attributes = video_hash.slice(*keys)
+    
+    video = Video.new(attributes)
+    video.subject_list.add(video_hash['subject'].strip.downcase)
```
Comment (DarryQueen) [inline app/models/video.rb]: If you're keeping the subjects lowercase, we'll have to find a way to titleize them later.

**Comment by DarryQueen** [comment]
Much of your `import_id` is repeated code. I would suggest keeping it there and refactoring code in `import_csv` to call the `import_id` method.

Also, tagging @jonathanshum to touch up the new video view.

**Comment by DarryQueen** [comment]
Added Naseem and Jonathan as reviewers, since I've mostly taken ownership of this pull request.

**Comment by nalnaji** [comment]
Looks good to me!
